### ISSUE TITLE
How to load model with load_state_dict ?

### ISSUE THREAD

**Issue by aymuos15** [ultralytics/ultralytics#18097]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

```
weights = f"{HOME}/runs/detect/train70/weights/best.pt"
model = YOLO(weights)
```

This works fine! But I am in a situation where I can not use this. 

My alternative which I found: https://github.com/ultralytics/yolov5/issues/1441#issuecomment-730305145

But this does not work. few layers throw mismatch error:
```

model = YOLO('yolo11n.yaml')
model.nc = 8 
model.model.load_state_dict(torch.load(weights)['model'].state_dict())

```

How do I go about this?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @aymuos15, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us debug it.

If this is a ❓ Question, please provide as much information as possible, including a detailed explanation of your use case and full error logs, as well as verifying compatibility with the latest `ultralytics` version. See the suggested upgrade command below.

For this specific inquiry regarding model loading using `load_state_dict` and its associated errors, it is important to ensure that:
- The model architecture (`yolo11n.yaml`) and the weights (`best.pt`) are compatible.
- Any custom modifications, such as altering `model.nc`, are correctly integrated and do not cause layer mismatches.

Feel free to share exact error messages or provide details about how these layers mismatch to help us gain further insight into the issue. An Ultralytics engineer will also assist you soon to provide more guidance 🙂.

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package, including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml), in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to ensure your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are currently passing. CI tests verify the correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by aymuos15** [comment]
`model.model.load_state_dict(torch.load(weights)['model'].state_dict(), strict=False)`

Even this throws the same error

**Comment by glenn-jocher** [comment]
@aymuos15 it seems the weight file might not completely match the model architecture. Instead, consider using the `load` method provided by the Ultralytics `YOLO` class: `model = YOLO('yolo11n.yaml').load(weights)`. This will handle any mismatched parameters more gracefully. For further details, refer to the [Ultralytics model documentation](https://docs.ultralytics.com/reference/engine/model/).

**Comment by Y-T-G** [comment]
Simply changing `model.nc` wouldn't work since the model is already created with the nc in the yaml. You will have to change nc in the yaml.

**Comment by Y-T-G** [comment]
Does your yaml's nc match with the weights?

**Comment by aymuos15** [comment]
This fixed my error. Thanks a lot! 

`Simply changing model.nc wouldn't work since the model is already created with the nc in the yaml. You will have to change nc in the yaml.`

I double checked now!

**Comment by glenn-jocher** [comment]
You're welcome! Glad it worked for you. Let us know if you have any further questions or run into any issues. Happy experimenting with YOLO!
